toolchain/check/testdata/class/generic/adapt.carbon

```diff
@@ -0,0 +1,912 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/class/generic/adapt.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/class/generic/adapt.carbon
+
+// --- adapt_specific_type.carbon
+
+library "[[@TEST_NAME]]";
+
+class C(T:! type) {
+  var x: T;
+}
+
+class Adapter {
+  adapt C(i32);
+}
+
+fn Access(a: Adapter) -> C(i32) {
+  // CHECK:STDERR: adapt_specific_type.carbon:[[@LINE+7]]:3: error: cannot implicitly convert from `i32` to `C(i32)` [ImplicitAsConversionFailure]
+  // CHECK:STDERR:   return (a as C(i32)).x;
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR: adapt_specific_type.carbon:[[@LINE+4]]:3: note: type `i32` does not implement interface `ImplicitAs(C(i32))` [MissingImplInMemberAccessNote]
+  // CHECK:STDERR:   return (a as C(i32)).x;
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  return (a as C(i32)).x;
+}
+
+// --- import_adapt_specific_type.carbon
+
+library "[[@TEST_NAME]]";
+
+import library "adapt_specific_type";
+
+fn ImportedAccess(a: Adapter) -> i32 {
+  return (a as C(i32)).x;
+}
+
+// --- fail_todo_extend_adapt_specific_type.carbon
+
+library "[[@TEST_NAME]]";
+
+class C(T:! type) {
+  var x: T;
+}
+
+class Adapter {
+  extend adapt C(i32);
+}
+
+fn Access(a: Adapter) -> C(i32) {
+  // TODO: This should presumably work, but the design doesn't say how yet.
+  // CHECK:STDERR: fail_todo_extend_adapt_specific_type.carbon:[[@LINE+13]]:3: error: cannot implicitly convert from `i32` to `C(i32)` [ImplicitAsConversionFailure]
+  // CHECK:STDERR:   return a.x;
+  // CHECK:STDERR:   ^~~~~~~~~~~
+  // CHECK:STDERR: fail_todo_extend_adapt_specific_type.carbon:[[@LINE+10]]:3: note: type `i32` does not implement interface `ImplicitAs(C(i32))` [MissingImplInMemberAccessNote]
+  // CHECK:STDERR:   return a.x;
+  // CHECK:STDERR:   ^~~~~~~~~~~
+  // CHECK:STDERR:
+  // CHECK:STDERR: fail_todo_extend_adapt_specific_type.carbon:[[@LINE+6]]:10: error: cannot implicitly convert from `Adapter` to `C(i32)` [ImplicitAsConversionFailure]
+  // CHECK:STDERR:   return a.x;
+  // CHECK:STDERR:          ^~~
+  // CHECK:STDERR: fail_todo_extend_adapt_specific_type.carbon:[[@LINE+3]]:10: note: type `Adapter` does not implement interface `ImplicitAs(C(i32))` [MissingImplInMemberAccessNote]
+  // CHECK:STDERR:   return a.x;
+  // CHECK:STDERR:          ^~~
+  return a.x;
+}
+
+// --- fail_todo_import_extend_adapt_specific_type.carbon
+
+library "[[@TEST_NAME]]";
+
+import library "adapt_specific_type";
+
+fn ImportedAccess(a: Adapter) -> i32 {
+  // TODO: This should presumably work, but the design doesn't say how yet.
```

Comment:
The comment makes it sound like this file fails, and it's named as though it fails, but I see no reported diagnostics.

Reply:
Yeah, oops, I got this test wrong too. Should be fixed now.